crates/ursa-gateway/src/main.rs

```diff
@@ -69,32 +79,112 @@ async fn main() -> Result<()> {
             let server_config = Arc::new(RwLock::new(gateway_config));
             let admin_config = Arc::clone(&server_config);
 
-            task::spawn(async move {
-                if let Err(e) = server::start(server_config, server_cache).await {
-                    error!("[Gateway server]: {e:?}");
-                };
-            });
-
-            task::spawn(async move {
-                if let Err(e) = admin::start(admin_config, admin_cache).await {
-                    error!("[Admin server]: {e:?}");
-                };
-            });
-
-            task::spawn(async move {
-                let duration_ms = Duration::from_millis(worker_ttl_interval);
-                info!("[Cache TTL Worker]: Interval: {duration_ms:?}");
-                loop {
-                    tokio::time::sleep(duration_ms).await;
-                    if let Err(e) = worker_tx.send(worker::cache::WorkerCacheCommand::TtlCleanUp) {
-                        error!("[Cache TTL Worker]: {e:?}");
+            let (shutdown_tx, shutdown_rx) = broadcast::channel(3);
+
+            let (server_worker, mut server_worker_signal_rx) = {
+                let (signal_tx, signal_rx) = mpsc::channel(1);
+                let worker = spawn(async move {
+                    if let Err(e) = server::start(server_config, server_cache, shutdown_rx).await {
+                        error!("[Server]: {e:?}");
+                        signal_tx.send(()).await.expect("Send signal successfully");
+                    };
+                    info!("Server stopped");
+                });
+                (worker, signal_rx)
+            };
+
+            let (admin_worker, mut admin_worker_signal_rx) = {
+                let shutdown_rx = shutdown_tx.subscribe();
+                let (signal_tx, signal_rx) = mpsc::channel(1);
+                let worker = spawn(async move {
+                    if let Err(e) = admin::start(admin_config, admin_cache, shutdown_rx).await {
+                        error!("[Admin server]: {e:?}");
+                        signal_tx.send(()).await.expect("Send signal successfully");
+                    };
+                    info!("Admin server stopped");
+                });
+                (worker, signal_rx)
+            };
+
+            let (ttl_cache_worker, mut ttl_cache_worker_signal_rx) = {
+                let mut shutdown_rx = shutdown_tx.subscribe();
+                let (signal_tx, signal_rx) = mpsc::channel(1);
+                let worker = spawn(async move {
```

Comment:
Are these tasks aborted as a result of the signal? Abort will help the task finish then cancel the task iirc

Reply:
I only the signal being sent out if an error to start. If I just send a sigint will they gracefully shutdown?

Reply:
> Are these tasks aborted as a result of the signal? Abort will help the task finish then cancel the task iirc

It's not only aborted but also waiting for tasks to be completed, in general, we don't want tasks to be ignored or lost.

Reply:
> I only the signal being sent out if an error to start. If I just send a sigint will they gracefully shutdown?

It covers two scenarios:
1. if one critical task error starts, notify others tasks to stop immediately.
2. on SIGNIT, notify all tasks to finish all alive requests and then stop.